JMMServer/Tasks/RecreateAllGroupsTask.cs

```diff
@@ -0,0 +1,373 @@
+
+        /// <summary>
+        /// Creates a new group that series will be put in during group re-calculation.
+        /// </summary>
+        /// <param name="session">The NHibernate session.</param>
+        /// <returns>The temporary <see cref="AnimeGroup"/>.</returns>
+        private AnimeGroup CreateTempAnimeGroup(ISessionWrapper session)
+        {
+            DateTime now = DateTime.Now;
+
+            var tempGroup = new AnimeGroup
+                {
+                    GroupName = "AAA Migrating Groups AAA",
+                    Description = "AAA Migrating Groups AAA",
+                    SortName = "AAA Migrating Groups AAA",
+                    DateTimeUpdated = now,
+                    DateTimeCreated = now
+                };
+
+            // We won't use AnimeGroupRepository.Save because we don't need to perform all the extra stuff since this is for temporary use only
+            session.Insert(tempGroup);
+
+            return tempGroup;
+        }
+
+        /// <summary>
+        /// Deletes the anime groups and user mappings as well as resetting group filters and moves all anime series into the specified group.
+        /// </summary>
+        /// <param name="session">The NHibernate session.</param>
+        /// <param name="tempGroupId">The ID of the temporary anime group to use for migration.</param>
+        private void ClearGroupsAndDependencies(ISessionWrapper session, int tempGroupId)
+        {
+            _log.Info("Removing existing AnimeGroups and resetting GroupFilters");
+
+            session.CreateSQLQuery(@"
+                DELETE FROM AnimeGroup_User;
+                DELETE FROM AnimeGroup WHERE AnimeGroupID <> :tempGroupId;
+
+                UPDATE AnimeSeries SET AnimeGroupID = :tempGroupId;
+                UPDATE GroupFilter SET GroupsIdsString = '{}';")
+                .SetInt32("tempGroupId", tempGroupId)
+                .ExecuteUpdate();
+
+            // We've deleted/modified all AnimeGroup_User/AnimeGroup/AnimeSeries/GroupFilter records, so update caches to reflect that
+            _animeGroupUserRepo.ClearCache();
+            _animeGroupRepo.ClearCache();
+            _animeSeriesRepo.ClearCache();
+            _groupFilterRepo.ClearCache();
+            _log.Info("AnimeGroups have been removed and GroupFilters have been reset");
+        }
+
+        private void UpdateAnimeSeriesContractsAndSave(ISessionWrapper session, IReadOnlyCollection<AnimeSeries> series)
+        {
+            _log.Info("Updating contracts for AnimeSeries");
+
+            // Update batches of AnimeSeries contracts in parallel. Each parallel branch requires it's own session since NHibernate sessions aren't thread safe.
+            // The reason we're doing this in parallel is because updating contacts does a reasonable amount of work (including LZ4 compression)
+            Parallel.ForEach(series.Batch(DefaultBatchSize), new ParallelOptions { MaxDegreeOfParallelism = 4 },

```

Comment:
I limited that because each task/thread is opening a new NHibernate session. I didn't want to overburden the database with concurrent queries, etc.
For example, if the CPU had 32 cores, then it could well open 32 connections to the database concurrently, and you'd end up running into IO contention, etc. at that point (for SQLite anyway, SQL Server, MySQL would probably be serving queries from data in memory).

I just chose 4 as an arbitrary number. I did test doubling it and didn't notice any difference on my machine (8 logical cores).